[GeoMechanicsApplication] Reuse Python script to run multiple stages

avdg81 opened the pull request:
**📝 Description**
The integration tests that called `run_stages` from the `test_helper.py` script now use the cleaned up Python script `run_multiple_stages.py`.

**🆕 Changelog**
Changes include:
- Removed functions `run_stages`, `get_stages`, `get_separated_directory_names`, and `get_separated_stages` from `test_helper.py`. Their usages have been replaced by `run_stages` from `run_multiple_stages.py`.
- Added a function that changes the present working directory to the given path as long as the corresponding context has not been left.
- Removed empty `setUp` and `tearDown` methods of the modified integration tests, since doing nothing is the default behavior. Thus no overrides are needed.
- The `run_stages` function from `run_multiple_stages.py` now returns a list of `GeoMechanicsAnalysis` instances, just like the replaced `run_stages` function from `test_helper.py` did.
- Removed a few empty `.mdpa` files and used `"input_type": "use_input_model_part"` to indicate that no model files need to be read for those stages.
- Some more test cases now read the output files to check the obtained results. The reason is that we can't use the database for any time step except for the last one.

avdg81 committed:
Reuse script that runs multiple geotechnical stages

- Added a context manager that sets the current working directory in the present context.
- Only read the `ProjectParameters.json` file when we are about to run that stage.
- Don't let the Python script that runs multiple stages modify the system's path.
- The 1D consolidation test now reuses the general run multiple stages function.

avdg81 committed:
Replaced `run_stages` from `test_helper.py`

All tests that used to call `run_stages` from `test_helper.py` now call `run_stages` from `run_multiple_stages.py`. In this way, we have removed some duplicated code.

avdg81 committed:
Fixed broken integration tests

After running the geomechanics analysis stages, return a list of run stages.

avdg81 committed:
Replaced two occurrences of `get_stages`

Prefer to use the Python script that runs multiple analysis stages.

avdg81 committed:
Removed empty `.mdpa` files

Rather use `"input_type": "use_input_model_part"`, since the model did not change over the stages.

avdg81 committed:
Removed last occurrence of `get_stages`

That required reading the analysis results from the output files.

avdg81 committed:
Removed `get_stages` from `test_helper.py`

This function was no longer used.

avdg81 committed:
Removed `setUp` and `tearDown` methods of touched Python scripts

WPK4FEM commented on applications/GeoMechanicsApplication/python_scripts/run_multiple_stages.py:
```diff
@@ -1,29 +1,28 @@
+            with open(filename, "r") as f:
+                stage_parameters = Kratos.Parameters(f.read())
+            result.append(analysis.GeoMechanicsAnalysis(model, stage_parameters))
+            result[-1].Run()
```
Apparently the -1 gives the last filled index of result. Tried to look this up for Python, but am probably using wrong keywords in my search.

avdg81 replied:
Yes, indexing a list with a negative number will start from the back. See also [this FAQ](https://docs.python.org/3/faq/programming.html#what-s-a-negative-index).

WPK4FEM commented on applications/GeoMechanicsApplication/python_scripts/run_multiple_stages.py:
```diff
@@ -1,29 +1,28 @@
+    result = []
+    with context_managers.set_cwd_to(project_path):
+        for filename in project_parameters_filenames:
+            with open(filename, "r") as f:
+                stage_parameters = Kratos.Parameters(f.read())
```
What happens if there are errors opening the file and stage_parameters is not properly set. Does it still append an analysis to result at line 22 and what are then the stage_parameters used.

avdg81 replied:
When opening the file yields an error, an exception will be raised. This exception will be caught by the enclosing `except` clause (if any). If there are none (like here), script execution will simply be terminated.

WPK4FEM commented on applications/GeoMechanicsApplication/python_scripts/run_multiple_stages.py:
```diff
@@ -1,29 +1,28 @@
+import sys
 
 import KratosMultiphysics as Kratos
 from KratosMultiphysics.GeoMechanicsApplication import *
```
Will this give a code smell that tells to be specific about what you import?

avdg81 replied:
Yes, you are entirely right. I generally prefer the `import ... [as ...]` syntax, since that won't pollute the global namespace. Also, it forces you to be more specific about which functions and/or classes you mean.

WPK4FEM commented on applications/GeoMechanicsApplication/tests/test_dirichlet_u.py:
```diff
@@ -29,18 +21,12 @@ def test_dirichlet_u(self):
+        output_data = []
+        reader = test_helper.GiDOutputFileReader()
+        for i in range(n_stages):
+            output_data.append(reader.read_output_from(os.path.join(project_path, f'dirichlet_u_stage{i+1}.post.res')))
```
Nice that the extra list with output file names is gone in this way.

WPK4FEM commented on applications/GeoMechanicsApplication/python_scripts/run_multiple_stages.py:
```diff
@@ -1,29 +1,28 @@
+import KratosMultiphysics.GeoMechanicsApplication.context_managers as context_managers
 
 
 def run_stages(project_path, n_stages):
```
It would be nice that at some point n_stages becomes an optional argument, such that on omission all stages present are executed. The n_stages argument could then be used to execute the first n stages.

avdg81 replied:
Agreed. Also, I can imagine that at some point we'd like to have the option to explicitly specify a set of filenames or a different filename pattern. But, indeed, such extensions need a bit more thought before attempting to implement them.

mcgicjn2 reviewed (dismissed):
This appears to do as expected and I like some of the improvements.

WPK4FEM commented on applications/GeoMechanicsApplication/tests/test_helper.py:
```diff
@@ -45,43 +45,6 @@ def run_kratos(file_path, model=None):
-    model = Kratos.Model()
-    stages = [analysis.GeoMechanicsAnalysis(model, stage_parameters) for stage_parameters in parameters_stages]
-    return stages
-
```
The removal of the duplicated code.

WPK4FEM commented on applications/GeoMechanicsApplication/tests/test_helper.py:
```diff
@@ -45,43 +45,6 @@ def run_kratos(file_path, model=None):
-    stages = [analysis.GeoMechanicsAnalysis(model, stage_parameters) for stage_parameters in parameters_stages]
-    return stages
-
 def get_separated_directory_names(project_path, n_stages):
```
This is the start of yet another "duplication" it is still used one test: test_element_lab.py test_element_lab/triaxial_comp_6n/Stage_{1,2} so its not far away to revise that test and also remove piece of scripting.

avdg81 replied:
Yes, I've seen that too. My initial intention was to address that later, but let me know if you feel that I'd better do it now. Thanks.

avdg81 replied:
As discussed, I have removed two more functions from `test_helper.py`.

WPK4FEM reviewed (commented):
Thank you for the clean up. Using context_managers.py makes the behaviour of the test suite more predictable as it solves the directory switching issues. I commented most on the changes in the functional part, the changes in the tests themselves as clear to me and a direct and predictable consequence of the changes in functionality.

avdg81 committed:
Removed yet another empty `.mdpa` file

avdg81 committed:
Removed a duplicated material parameters file

avdg81 committed:
Read calculated stresses from the output files

avdg81 committed:
Reuse `run_stages` from `run_multiple_stages.py`

Moved a few files and removed two subdirectories.

Also improved the assertions: make sure that the stress vector results have the correct sizes.

avdg81 committed:
Removed commented out element types and condition types

avdg81 committed:
Removed two unused functions from `test_helper.py`

avdg81 committed:
Removed redundant overrides of `setUp` and `tearDown`

avdg81 committed:
Removed an `import` statement that appeared to be redundant
